File: src/CodexQuotaRail.Core/Quotas/QuotaNormalizer.cs
namespace CodexQuotaRail.Core.Quotas;

public static class QuotaNormalizer
{
    public static QuotaDisplayState Normalize(RawQuotaSnapshot snapshot)
    {
        var windows = new[] { snapshot.Primary, snapshot.Secondary }
            .Where(window => window is not null)
            .Select(window => NormalizeWindow(window!))
            .ToArray();

        return new QuotaDisplayState(windows, QuotaConnectionState.Live, snapshot.ReceivedAt, null);
    }

    public static QuotaWindowDisplay NormalizeWindow(RawQuotaWindow source)
    {
        if (source.IsUnlimited)
        {
            return new(source.Label, null, Duration(source), Reset(source), QuotaWindowState.Unlimited);
        }

        if (source.UsedPercent is null)
        {
            return new(source.Label, null, Duration(source), Reset(source), QuotaWindowState.Unavailable);
        }

        var available = Math.Clamp(100 - source.UsedPercent.Value, 0, 100);
        var state = available switch
        {
            0 => QuotaWindowState.Exhausted,
            <= 20 => QuotaWindowState.Critical,
            <= 50 => QuotaWindowState.Notice,
            _ => QuotaWindowState.Healthy
        };

        return new(source.Label, available, Duration(source), Reset(source), state);
    }

    private static TimeSpan? Duration(RawQuotaWindow source) =>
        source.WindowDurationMins is long minutes ? TimeSpan.FromMinutes(minutes) : null;

    private static DateTimeOffset? Reset(RawQuotaWindow source) =>
        source.ResetsAtUnixSeconds is long seconds ? DateTimeOffset.FromUnixTimeSeconds(seconds) : null;
}

File: macos/Packages/CodexQuotaKit/Sources/CodexQuotaKit/QuotaNormalizer.swift
import Foundation

public enum QuotaNormalizer {
    public static func availablePercent(from usedPercent: Int) -> Int {
        min(100, max(0, 100 - usedPercent))
    }

    public static func normalize(_ snapshot: RawQuotaSnapshot) -> QuotaDisplayState {
        let windows = [snapshot.primary, snapshot.secondary].compactMap(mapWindow)
        return QuotaDisplayState(
            windows: windows,
            connection: .live,
            updatedAt: snapshot.receivedAt,
            message: nil)
    }

    private static func mapWindow(_ source: RawQuotaWindow?) -> QuotaWindowDisplay? {
        guard let source else {
            return nil
        }
        if source.isUnlimited {
            return QuotaWindowDisplay(
                label: source.label,
                availablePercent: nil,
                windowDurationMinutes: source.windowDurationMinutes,
                resetsAt: source.resetsAt,
                state: .unlimited)
        }
        guard let usedPercent = source.usedPercent else {
            return QuotaWindowDisplay(
                label: source.label,
                availablePercent: nil,
                windowDurationMinutes: source.windowDurationMinutes,
                resetsAt: source.resetsAt,
                state: .unavailable)
        }
        let available = availablePercent(from: usedPercent)
        return QuotaWindowDisplay(
            label: source.label,
            availablePercent: available,
            windowDurationMinutes: source.windowDurationMinutes,
            resetsAt: source.resetsAt,
            state: state(for: available))
    }

    private static func state(for availablePercent: Int) -> QuotaWindowState {
        if availablePercent == 0 {
            return .exhausted
        }
        if availablePercent <= 20 {
            return .critical
        }
        if availablePercent <= 50 {
            return .notice
        }
        return .healthy
    }
}
